src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs

```diff
@@ -12,30 +12,40 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
     using System.Threading.Tasks;
     using Xunit;
 
-    public abstract class DiagnosticsTests : IDisposable
+    public abstract class DiagnosticsTests
     {
-        protected ConcurrentQueue<(string eventName, object payload, Activity activity)> events;
-        protected FakeDiagnosticListener listener;
-        protected IDisposable subscription;
-        protected const int maxWaitSec = 10;
+        protected const int MaxWaitSec = 10;
 
-        private bool disposed = false;
+        protected ConcurrentQueue<(string eventName, object payload, Activity activity)> CreateEventQueue() =>
+            new ConcurrentQueue<(string eventName, object payload, Activity activity)>();
 
-        internal DiagnosticsTests()
-        {
-            this.events = new ConcurrentQueue<(string eventName, object payload, Activity activity)>();
-            this.listener = new FakeDiagnosticListener(kvp =>
+        protected FakeDiagnosticListener CreateEventListener(string entityName, ConcurrentQueue<(string eventName, object payload, Activity activity)> eventQueue) =>
```

Comment:
What about disabling parallelization for tests that are using EventListener? Shouldn't make runs much slower as we don't have too many of them but might make things more stable.

Reply:
To accomplish that you can put tests in common collection with `CollectionDefinition(DisableParallelization = true)]`

Reply:
The problem is that ANY activity in the Service Bus library causes diagnostic events and activities to flow.   You'd essentially have to put every single Live test in that collection, removing parallelization.   That was the original position for the tests and there were still some intermittent failures because of diagnostic activities crossing test boundaries.  Adding isolation was a focus to remove those inconsistencies, and had the nice side-effect of allowing parallel tests - which took the overall time from ~21 minutes to ~3 minutes for a full live test run.

The only way to isolate these that I found is to scope a specific listener to only the entities used by a specific test.   That said, I've limited hands-on with the framework diagnostics source/listeners, so you may have a trick or three up your sleeve that I'm not aware of.   Happy to discuss further if you've got thoughts.

Reply:
I was under the impression that there only parts of live tests use EventListener. If all of them do then you are right and there is no point in limiting the parallelism.

Reply:
I looked again, there is a ton of live tests that are not using the diagnostic listener. Running all of them in parallel if completely fine, `[CollectionDefinition(DisableParallelization = true)]` would force diagnostic tests to run after all other tests were run in parallel.

Reply:
Are we talking about two different things here?   I'm noting that the Service Bus library being tested is producing diagnostic events that flow through the listener.   Even though only the diagnostics family of tests make use of the listener, without the filtering that I put in to keep a listener focused on activities for a specific entity, the listener would be grabbing data from all activities happening in the other tests.

I feel like I'm missing your point here.  What are you proposing, exactly?   I thought it was removing the entity filtering, but ....

Reply:
> the listener would be grabbing data from all activities happening in the other tests.

Why if there would be no other tests running in parallel? 

>  What are you proposing, exactly? I thought it was removing the entity filtering, but ...

Moving diagnostics test into a collection so they won't run in parallel with anything else or each other and removing the filtering 

Maybe I'm missing something :)

Reply:
I think that you may be.   The listener is an instance, not shared among the tests.   Each instance is filtering diagnostic information to only the entities that it's owner cares about.   However, the source of diagnostics is global.  Any operation that uses the Service Bus client library potentially produces diagnostic information.   That information is pushed to every listener instance that is an observer, with a predicate called to allow filtering.    The underlying problem wasn't that listeners were being shared by the tests, it was that each listener is observing the global diagnostics and was seeing information produced by other callers in all other tests - whether they have a listener or not.  The producer is still creating diagnostics for them and sending it to everyone.

For a fun illustration, lets say that every 3 seconds, a different person smacks me in the head.  You and Alex are watching me get smacked.  You only care about people with blond hair that smack me but you see every single smack because there's only one me, and I'm being beaten in public.    Your only choice is to teach yourself to ignore it when someone who isn't blond hits me.

Reply:
Pavel and I discussed offline;  I wasn't fully understanding his suggestion.   Going to give it a look and see what the net effect is on overall stability.

Reply:
After some investigation, this seemed to work out well.   I've converted the approach to remove filtering by entity and, instead, group diagnostic tests into a collection to be run as a single unit of execution.  They'll run in isolation with each diagnostic test done serially to ensure that there are no other consumers of the client generating diagnostics information that may interfere.

Note that I did keep the logic in the `DiagnosticTests` base class that identifies mismatches between the entity a test is interested in and activities for a different target entity.  Instead of filtering, it now emits a log entry.  This information was incredibly useful in diagnosing the test instability, so I believe it worth the additional complexity and time to execute.